### PR TITLE
[List Item] List Item Prototypes

### PR DESCRIPTION
**Pull request by andrewoverton**
This PR includes two new ideas for approaching List Items—`AutoLayoutListItemCell` and `ManualLayoutListItemCell`, to be considered alongside @yarneo’s `CollectionViewListCell`.

My goal in these two classes was to try to be able to represent as much of what’s outlined in [the spec](https://material.io/design/components/lists.html) as possible with one cell class. I think by that metric what I have is a good start. These cells allow for any arbitrary configuration of leading and trailing supporting visuals, be they controls, image views, labels, you name it. However, there is inevitable overhead that comes with altering the view hierarchy in a reusable cell to allow for this flexibility. If we agree that this is unacceptable, something closer to what Yarden originally did in `CollectionViewListCell` is a much better option. Such an MVP would simply have a leading image, a trailing image, some title text, and some detail text. On the other hand, if it is decided that the pros of increased configurability outweigh the cons associated with it, then one of these classes I’ve been working on will be worth keeping around and improving upon. Personally, of these two classes, I prefer `ManualLayoutListItemCell`. With this choice, the performance costs of altering the view hierarchy are offset by the lack of Auto Layout constraints to reconcile.

Should we decide to continue developing `ManualLayoutListItemCell`, here is an incomplete list of things that still need addressing: 
* Accessibility in general
* RTL support
* Some general polishing—pixel auditing, etc.
* The logic backing several property setters in `ManualLayoutListItemCell` still needs to be fleshed out. I’m referring to things like `centerLeadingViewVertically`, `automaticallySetTextOffset`, and `textOffset`. These are speculative properties anyway, which might not survive an API review.

Lastly, here are some low-ish priority open questions/things I’ve been thinking about over the last week:
* Do we want any kind of built-in cell divider? The spec shows cell dividers that AFAIK are not provided by UIKit. 
    * If we do want to include one, how customizable should it be?
    * Should it be a supplementary view or should it be in the cell itself?
* The spec mentions support for swiping left or right to perform actions/access underlying controls. Do we want to support this immediately? Save it for a second iteration? Get to it if we have time? Last Friday I began prototyping an entirely separate `SwipeRevealer` class that could make this possible. It achieves the functionality with some gesture recognizers and some UIView properties for `hostView`, `underlyingView`, and `snapshotView`. More on that to come.
* The spec mentions support for dragging to rearrange. This immediately strikes me as work for whoever owns the collection view. However, we could consider adding an editing mode to put the cells in when this kind of things happen. Should we consider this? Or should it be up to the view controller to update the leadingView to a “reorder control” view of some sort during editing?
* The spec mentions behavioral changes in different size environments. 
    * Example 1: Regular lists on portrait iPhones become image lists on landscape iPads.
        * Do we want to implement this in our V1 component? It increases the complexity substantially. Should it be its own component?
    * Example 2: Vertically organized labels within a single list item on portrait iPhones become horizontally organized (columnar) labels on landscape iPads.
        * Do we want to implement this in our V1 component? It increases the complexity, but could be made possible with a `BOOL` property and some attention paid to size classes.

### CODE REVIEW

File: components/List/examples/AutoLayoutListCellExample.m [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    (collectionView.adjustedContentInset.left + collectionView.adjustedContentInset.right);
+  }
+#endif
+  cell.cellWidth = cellWidth;
+
+  cell.typographyScheme = _typographyScheme;
+  cell.mdc_adjustsFontForContentSizeCategory = YES;
+
+  if (indexPath.item % 3 == 0) {
+    UIImageView *leadingView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 70, 70)];
```
Comment (ianegordon) [inline components/List/examples/AutoLayoutListCellExample.m:108]: Adding a view here (and removing them below) may have performance implications.

File: components/List/examples/AutoLayoutListCellExample.m [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    cell.overlineText = nil;
+    cell.titleText = nil;
+    cell.detailText = nil;
+  }
+
+  if (indexPath.item % 2 == 0) {
+    cell.automaticallySetTextOffset = YES;
+  }
+
+  if (indexPath.item == 1) {
```
Comment (ianegordon) [inline components/List/examples/AutoLayoutListCellExample.m:146]: Merge this block with the .item =1 block above?

File: components/List/examples/supplemental/AutoLayoutListItemCell.m [hunk 3]
```diff
@@ -0,0 +1,803 @@
...
+    self.trailingContainerTrailingConstraint.constant = 0;
+    self.trailingContainerTopConstraint.constant = 0;
+    self.trailingContainerBottomConstraint.constant = 0;
+  }
+
+  [self setNeedsLayout];
+}
+
+-(void)setOverlineText:(NSString *)overlineText {
+  if (overlineText == _overlineText) {
```
Comment (ianegordon) [inline components/List/examples/supplemental/AutoLayoutListItemCell.m:630]: nit: || isEqualToString:

File: components/List/examples/supplemental/AutoLayoutListItemCell.m [hunk 4]
```diff
@@ -0,0 +1,803 @@
...
+    return;
+  }
+  _overlineText = overlineText;
```
Comment (ianegordon) [inline components/List/examples/supplemental/AutoLayoutListItemCell.m:633]: copy parameters with Mutable subclasses.
_overlineText = [overlineText copy];

File: components/List/examples/supplemental/AutoLayoutListItemCell.m [hunk 5]
```diff
@@ -0,0 +1,803 @@
...
+                               scaledForDynamicType:_mdc_adjustsFontForContentSizeCategory];
+  }
+  self.overlineLabel.font = overlineFont;
+  self.titleLabel.font = titleFont;
+  self.detailLabel.font = detailFont;
+  [self setNeedsLayout];
+}
+
+- (void)layoutSubviews {
+  [super layoutSubviews];
```
Comment (ianegordon) [inline components/List/examples/supplemental/AutoLayoutListItemCell.m:789]: Remove?

File: components/List/examples/supplemental/ManualLayoutListItemCell.m [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+@implementation ManualLayoutListItemCell
+@synthesize mdc_adjustsFontForContentSizeCategory = _mdc_adjustsFontForContentSizeCategory;
+
+#pragma mark Object Lifecycle
+
+-(instancetype)initWithFrame:(CGRect)frame {
+  self = [super initWithFrame:frame];
+  if (self) {
+    [self commonInit];
+    return self;
```
Comment (ianegordon) [inline components/List/examples/supplemental/ManualLayoutListItemCell.m:52]: Our init methods usually follow the following pattern:
self = [super init...];
if (self) {
  [self commonClassNameInit];
}
return self;

File: components/List/examples/supplemental/ManualLayoutListItemCell.m [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+@implementation ManualLayoutListItemCell
+@synthesize mdc_adjustsFontForContentSizeCategory = _mdc_adjustsFontForContentSizeCategory;
+
+#pragma mark Object Lifecycle
+
+-(instancetype)initWithFrame:(CGRect)frame {
+  self = [super initWithFrame:frame];
+  if (self) {
+    [self commonInit];
```
Comment (ianegordon) [inline components/List/examples/supplemental/ManualLayoutListItemCell.m:51]: note: This name will collide in framework code.  Embed class name in method name, commonManualLayoutListItemCellInit

File: components/List/examples/supplemental/ManualLayoutListItemCell.m [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+  return
+  self.overlineLabel.text.length > 0 ||
+  self.titleLabel.text.length > 0 ||
+  self.detailLabel.text.length > 0;
+}
+
+#pragma mark - Typography/Dynamic Type Support
+
+- (MDCTypographyScheme *)defaultTypographyScheme {
+  return [MDCTypographyScheme new];
```
Comment (ianegordon) [inline components/List/examples/supplemental/ManualLayoutListItemCell.m:297]: nit: We tend to prefer alloc-init over new.

File: components/List/examples/supplemental/AutoLayoutListItemCell.m [hunk 9]
```diff
@@ -0,0 +1,803 @@
...
+}
+
+#pragma mark UICollectionViewCell Overrides
+
+-(void)prepareForReuse {
+  [super prepareForReuse];
+  self.overlineText = nil;
+  self.titleText = nil;
+  self.detailText = nil;
+  self.leadingView = nil;
```
Comment (ianegordon) [inline components/List/examples/supplemental/AutoLayoutListItemCell.m:128]: If there are performance concerns, this is where I would investigate hidden, over removing the view from the hierarchy.

**Comment by jverkoey** [comment]
@ianegordon @andrewoverton do we feel like we can pursue landing this PR into our examples?

**Comment by andrewoverton** [comment]
@jverkoey I was thinking I'd just close this one and open another in the next few days for the API review. The implementation won't be completely fleshed out, but my understanding is that's okay for an API review.

**Comment by jverkoey** [comment]
Ok! Sounds fair.

**Pull request closed without merging**